state_tracking.go

```diff
@@ -0,0 +1,181 @@
+package kapacitor
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/influxdata/kapacitor/models"
+	"github.com/influxdata/kapacitor/pipeline"
+	"github.com/influxdata/kapacitor/tick/ast"
+	"github.com/influxdata/kapacitor/tick/stateful"
+)
+
+type stateTrackingGrouper interface {
```

Comment:
I like how they are combined and share the common implementation. To me the name grouper was confusing at first since I thought it was responsible for grouping the data not tracking each group. Maybe this should be called `stateTracker` and the method is called `track` instead of `pointResult`? That way the `tracker` `track`s which is how Go interfaces should read. Then `newGrouper` would be `newTracker` etc.

Thoughts? Not super important as like you said naming is easy to change later.